server/src/main/java/com/objectcomputing/checkins/services/opportunities/OpportunitiesServicesImpl.java

```diff
@@ -0,0 +1,98 @@
+            } else if(memberRepo.findById(memberId).isEmpty()){
+                throw new BadArgException(String.format("Member %s doesn't exists", memberId));
+            } else if(surSubDate.isBefore(LocalDate.EPOCH) || surSubDate.isAfter(LocalDate.MAX)) {
+                throw new BadArgException(String.format("Invalid date for opportunities submission date %s",memberId));
+            }
+            opportunitiesResponseRet = opportunitiesResponseRepo.save(opportunitiesResponse);
+        }
+        return opportunitiesResponseRet ;
+    }
+
+    public Set<Opportunities> readAll() {
+        return opportunitiesResponseRepo.findAll();
+    }
+
+    @Override
+    public Opportunities update(Opportunities opportunitiesResponse) {
+        final boolean isAdmin = currentUserServices.isAdmin();
+        permissionsValidation.validatePermissions(!isAdmin, "User is unauthorized to do this operation");
+        Opportunities opportunitiesResponseRet = null;
+        if(opportunitiesResponse!=null){
+            final UUID id = opportunitiesResponse.getId();
+            final UUID memberId = opportunitiesResponse.getSubmittedBy();
+            LocalDate surSubDate = opportunitiesResponse.getSubmittedOn();
+            if(id==null||opportunitiesResponseRepo.findById(id).isEmpty()){
+                throw new BadArgException(String.format("Unable to find opportunities record with id %s", opportunitiesResponse.getId()));
+            }else if(memberRepo.findById(memberId).isEmpty()){
+                throw new BadArgException(String.format("Member %s doesn't exist", memberId));
+            } else if(memberId==null) {
+                throw new BadArgException(String.format("Invalid opportunities %s", opportunitiesResponse));
+            } else if(surSubDate.isBefore(LocalDate.EPOCH) || surSubDate.isAfter(LocalDate.MAX)) {
+                throw new BadArgException(String.format("Invalid date for opportunities submission date %s",memberId));
+            }
+
+            opportunitiesResponseRet = opportunitiesResponseRepo.update(opportunitiesResponse);
+        }
+        return opportunitiesResponseRet;
+    }
+
+    @Override
+    public void delete(@NotNull UUID id) {
+        final boolean isAdmin = currentUserServices.isAdmin();
+        permissionsValidation.validatePermissions(!isAdmin, "User is unauthorized to do this operation");
+        opportunitiesResponseRepo.deleteById(id);
+    }
+
+    @Override
+    public Set<Opportunities> findByFields(String name, String description) {
+        Set<Opportunities> opportunitiesResponse = new HashSet<>();
+        opportunitiesResponseRepo.findAll().forEach(opportunitiesResponse::add);
+        if(name!=null){
+            opportunitiesResponse.addAll(opportunitiesResponseRepo.findByName(name));
+        }
+        else if(description!=null){
+            opportunitiesResponse.addAll(opportunitiesResponseRepo.findByDescription(description));
+        }
+        return opportunitiesResponse;
```

Comment:
You are adding all opportunities here regardless of what the search string is. Did you mean to use the retain all feature in your conditional statements instead?

Reply:
@warnerj-oci I think you are probably right (that @Maroju100 meant to use retailAll here)...but the retainAll approach to these find by fields methods is also not desirable. We should be using the "is null or..." approach in our queries to push that to the database and not put us in the position of loading everything into memory and then filtering. Also, it allows for a single query rather than subsequent queries to the database to filter down.